iati/core/rulesets.py

```diff
@@ -1,45 +1,841 @@
+
+        try:
+            re.compile(self.regex)
+        except sre_constants.error:
+            raise ValueError
+        if self.regex == '':
+            raise ValueError
+
+    def __str__(self):
+        """Return string stating what RuleRegexNoMatches is checking."""
+        if len(self.paths) == 1:
+            return 'Each `{self.paths[0]}` within each `{self.context}` must not match the regular expression `{self.regex}`.'.format(**locals())
+        return 'Each instance of `{0}` within each `{self.context}` must not match the regular expression `{self.regex}`.'.format('` and `'.join(self.paths), **locals())
+
+    def _check_against_Rule(self, context_element):
+        """Assert that no text of the given `paths` matches the `regex` value.
+
+        Args:
+            context_element (Element): An XML Element.
+
+        Returns:
+            bool: Return `True` when the given `path` text does not match the given `regex`.
+                  Return `False` when the given `path` text matched the given `regex`.
+
+        """
+        pattern = re.compile(self.regex)
+
+        for path in self.paths:
+            strings_to_check = self._extract_text_from_element_or_attribute(context_element, path)
+            for string_to_check in strings_to_check:
+                if pattern.search(string_to_check):
+                    return False
+        return True
+
+
+class RuleStartsWith(Rule):
+    """Representation of a Rule that checks that the prefixing text of each text value for `path` matches the `start` text value."""
+
+    def __init__(self, context, case):
+        """Initialise a `startswith` Rule."""
+        self.name = 'startswith'
+
+        super(RuleStartsWith, self).__init__(context, case)
+
+    def __str__(self):
+        """Return string stating what RuleStartsWith is checking."""
+        if len(self.paths) == 1:
+            return 'Each `{self.paths[0]}` within each `{self.context}` must start with the value present at `{self.start}`.'.format(**locals())
+        return 'Each instance of `{0}` within each `{self.context}` must start with the value present at `{self.start}`.'.format('` and `'.join(self.paths), **locals())
+
+    def _normalize_xpaths(self):
+        """Normalize xpaths by combining them with `context`."""
+        super(RuleStartsWith, self)._normalize_xpaths()
+
+        self.normalized_paths.append(self._normalize_xpath(self.start))
+
+    def _check_against_Rule(self, context_element):
+        """Assert that the prefixing text of all given `paths` starts with the text of `start`.
+
+        Args:
+            context_element (Element): An XML Element.
+
+        Returns:
+            bool: Return `True` when the `path` text starts with the text value of `start`.
+                  Return `False` when the `path` text does not start with the text value of `start`.
+
+        """
+        prefix = self._extract_text_from_element_or_attribute(context_element, self.start)[0]
```

Comment:
# NON-TINY ISSUE

The definition of `startswith` specifies a singular element matched by `start`. As such, if there are multiple matched elements, this should lead to an error or failure or something.

The current implementation also looks like it'd cause problems if nothing is matched by the `start` value.

Reply:
💩

Reply:
done in separate pull request